### PR TITLE
[ONNX] Raise exception for mixed precision input for BatchNormalization

### LINKED ISSUE #72494: onnx.export() of Batch1dNorm generates invalid ONNX under torch.cuda.amp.autocast()
### 🐛 Describe the bug

onnx.export() of Batch1dNorm generates invalid ONNX under torch.cuda.amp.autocast().
Here is a small repro case: 

## Repro 
```
import torch
import torch.nn as nn
import onnx


class BatchNorm1dNoAutoCast(nn.BatchNorm1d):
    def init(self, num_features, **kwargs):
        nn.BatchNorm1d.init(self, num_features, **kwargs)

    def forward(self, x):
        if False:  # torch.onnx.is_in_onnx_export():
            ret = nn.BatchNorm1d.forward(self, x.to(torch.float)).to(x.dtype)
        else:
            ret = nn.BatchNorm1d.forward(self, x)
        return ret


mod = BatchNorm1dNoAutoCast(5).to(device="cuda").eval()

data = torch.rand(300, 5, device="cuda").half()
with torch.inference_mode(), torch.cuda.amp.autocast():
    torch.onnx.export(mod,
                      (data),
                      "a.onnx",
                      opset_version=14,
                      verbose=True
                      )

onnx_model = onnx.load("a.onnx")
onnx.checker.check_model(onnx_model, full_check=True)
```
##

## Error
onnx.onnx_cpp2py_export.shape_inference.InferenceError: [ShapeInferenceError] (op_type:BatchNormalization, node name: BatchNormalization_0): scale has inconsistent type tensor(float)             
##
Uncomment the code under torch.onnx.is_in_onnx_export() condition to get correct ONNX exported.

### Versions

Here.  I have also verified the bug manifests with latest 1.11a (NVIDIA pytocrh 22.01).
Seems to happen with all latest upsets : 13, 14, 15.

Collecting environment information...
PyTorch version: 1.10.1+cu113
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.3 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.21.4
Libc version: glibc-2.31

Python version: 3.8.10 (default, Sep 28 2021, 16:10:42)  [GCC 9.3.0] (64-bit runtime)
Python platform: Linux-5.11.0-37-generic-x86_64-with-glibc2.29
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration:
GPU 0: Quadro GV100
GPU 1: Quadro GV100
GPU 2: NVIDIA GeForce GT 710

Nvidia driver version: 470.57.02
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.3.0
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.20.3
[pip3] pytorch-lightning==1.5.9
[pip3] torch==1.10.1+cu113
[pip3] torch-stft==0.1.4
[pip3] torchaudio==0.10.1+cu113
[pip3] torchmetrics==0.6.0
[pip3] torchvision==0.11.2+cu113
[conda] Could not collect

cc @ngimel @mcarilli @ptrblck

### PR DESCRIPTION
**Pull request by thiagocrepaldi**
Fixes #72494

### CODE REVIEW

File: torch/onnx/symbolic_opset14.py [hunk 1]
```diff
@@ -40,6 +40,13 @@ def reshape(g, self, shape):
 
 @parse_args("v", "v", "v", "v", "v", "i", "f", "f", "i")
 def batch_norm(g, input, weight, bias, running_mean, running_var, training, momentum, eps, cudnn_enabled):
+
+    if torch.is_autocast_enabled() and \
```
Comment (garymm) [inline torch/onnx/symbolic_opset14.py:44]: Why condition on autocast?

File: torch/onnx/symbolic_helper.py [hunk 2]
```diff
@@ -970,6 +970,12 @@ def requantize_bias_helper(g, bias, input_scale, weight_scale):
                   to_i=torch.onnx.TensorProtoDataType.INT32)
     return g.op("prim::TupleConstruct", q_bias, bias_scale, bias_zero_point)
 
+def check_args_have_same_dtype(args):
+    assert args is not None and len(args) > 0
+    base_dtype = args[0].type().scalarType()
+    has_same_dtype = all([elem.type().scalarType() == base_dtype for elem in args])
```
Comment (garymm) [inline torch/onnx/symbolic_helper.py:976]: `all` can consume the iterator directly, no need to construct a list:
```suggestion
    has_same_dtype = all(elem.type().scalarType() == base_dtype for elem in args)
```

File: torch/onnx/symbolic_helper.py [hunk 3]
```diff
@@ -970,6 +970,12 @@ def requantize_bias_helper(g, bias, input_scale, weight_scale):
                   to_i=torch.onnx.TensorProtoDataType.INT32)
     return g.op("prim::TupleConstruct", q_bias, bias_scale, bias_zero_point)
 
+def check_args_have_same_dtype(args):
+    assert args is not None and len(args) > 0
```
Comment (garymm) [inline torch/onnx/symbolic_helper.py:974]: Equivalent, shorter, and more pythonic
```suggestion
    assert args
```

File: torch/onnx/symbolic_helper.py [hunk 4]
```diff
@@ -970,6 +970,12 @@ def requantize_bias_helper(g, bias, input_scale, weight_scale):
                   to_i=torch.onnx.TensorProtoDataType.INT32)
     return g.op("prim::TupleConstruct", q_bias, bias_scale, bias_zero_point)
 
+def check_args_have_same_dtype(args):
```
Comment (garymm) [inline torch/onnx/symbolic_helper.py:973]: Equivalent and shorter name, also makes it obvious what the return type is from the name:
```suggestion
def args_have_same_dtype(args):
```

File: torch/onnx/symbolic_opset14.py [hunk 5]
```diff
@@ -40,6 +40,13 @@ def reshape(g, self, shape):
...
+            not check_args_have_same_dtype([input, weight, bias, running_mean, running_var]) and \
+            sym_help._export_onnx_opset_version < 15:
+        return sym_help._onnx_opset_unsupported_detailed("BatchNormalization", 14, 15,
+                                                         "Tensors with mixed precision are not supported")
```
Comment (garymm) [inline torch/onnx/symbolic_opset14.py:48]: Let's try to make this error message more actionable. Maybe something like:

"All input tensors must have the same dtype. Try turning off autocast or exporting to opset_version=15"

**Review by garymm — COMMENTED** [review]
Please add a test to ensure the autocast + batchnorm actually works with opset_version=15.

File: torch/onnx/symbolic_opset14.py (hunk 1 above)
Comment (thiagocrepaldi) [reply torch/onnx/symbolic_opset14.py:44]: To minimize this change surface as type check is already done when autocast is not enabled.

File: torch/onnx/symbolic_helper.py [hunk 6]
```diff
@@ -970,6 +970,12 @@ def requantize_bias_helper(g, bias, input_scale, weight_scale):
                   to_i=torch.onnx.TensorProtoDataType.INT32)
     return g.op("prim::TupleConstruct", q_bias, bias_scale, bias_zero_point)
 
+def args_have_same_dtype(args):
+    assert args
+    base_dtype = args[0].type().scalarType()
+    has_same_dtype = all(elem.type().scalarType() == base_dtype for elem in args)
```
Comment (garymm) [inline torch/onnx/symbolic_helper.py:976]:
```suggestion
    return all(elem.type().scalarType() == base_dtype for elem in args)
```

**Review by garymm — CHANGES_REQUESTED** [review]
I guess you missed this last time, so repeating:
Please add a test to ensure the autocast + batchnorm actually works with opset_version=15.

File: torch/onnx/symbolic_opset14.py (hunk 1 above)
Comment (thiagocrepaldi) [reply torch/onnx/symbolic_opset14.py:44]: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/templates/TensorMethods.cpp#L6 and https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/TensorCompare.cpp are a generic pieces for the tensor implementation, but other places such as https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Blas.cpp#L121, https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/Blas.cpp#L381 also implement their own checks (there are other spots too).

**Comment by thiagocrepaldi** [comment]
> I guess you missed this last time, so repeating: Please add a test to ensure the autocast + batchnorm actually works with opset_version=15.

done

File: test/onnx/test_operators.py [hunk 7]
```diff
@@ -1055,6 +1055,20 @@ def forward(self, x):
                         input_names=["x"],
                         dynamic_axes={"x": {0: "dim_0"}})
 
+    def test_batch_norm(self):
+        class BatchNorm1dExport(nn.BatchNorm1d):
```
Comment (garymm) [inline test/onnx/test_operators.py:1059]: This class appears to be a no-op. Why not just use nn.BatchNorm1d directly?

File: test/onnx/test_operators.py [hunk 8]
```diff
@@ -1055,6 +1055,20 @@ def forward(self, x):
                         input_names=["x"],
                         dynamic_axes={"x": {0: "dim_0"}})
 
+    def test_batch_norm(self):
```
Comment (garymm) [inline test/onnx/test_operators.py:1058]: Can we add this the onnxruntime test instead?
I really don't want to add more expect tests unless we don't have a better idea. In this case, I think a. expect file is not useful.

File: torch/onnx/symbolic_opset14.py [hunk 9]
```diff
@@ -40,6 +40,13 @@ def reshape(g, self, shape):
 
 @parse_args("v", "v", "v", "v", "v", "i", "f", "f", "i")
 def batch_norm(g, input, weight, bias, running_mean, running_var, training, momentum, eps, cudnn_enabled):
+    if torch.is_autocast_enabled() and \
```
Comment (garymm) [inline torch/onnx/symbolic_opset14.py:43]:
```suggestion
   # PyTorch handles type checking unless autocast is enabled.
    if torch.is_autocast_enabled() and \
```

File: torch/onnx/symbolic_opset9.py [hunk 10]
```diff
@@ -1358,6 +1363,14 @@ def conv_transpose3d(g, input, weight, bias, stride, padding, output_padding, gr
 @parse_args("v", "v", "v", "v", "v", "i", "f", "f", "i")
 def batch_norm(g, input, weight, bias, running_mean, running_var, training, momentum, eps, cudnn_enabled):
     sym_help.check_training_mode(training, "batch_norm")
+
```
Comment (garymm) [inline torch/onnx/symbolic_opset9.py:1366]: please add same comment here.

File: torch/onnx/symbolic_opset14.py (hunk 1 above)
Comment (garymm) [reply torch/onnx/symbolic_opset14.py:43]: Thanks for explaining.

**Review by garymm — CHANGES_REQUESTED** [review]

File: test/onnx/test_operators.py (hunk 8 above)
Comment (thiagocrepaldi) [reply test/onnx/test_operators.py:1058]: There are several tests for batch norm for onnxruntime that does the same thing.

https://github.com/pytorch/pytorch/blob/master/test/onnx/test_pytorch_onnx_onnxruntime.py#L3563-L3610

Comment (garymm) [reply test/onnx/test_operators.py]: Please modify that file to test specifically for the mixed type case.

File: test/onnx/test_pytorch_onnx_onnxruntime.py [hunk 11]
```diff
@@ -3560,6 +3560,17 @@ def test_layer_norm(self):
         x = torch.randn(20, 5, 10, 10)
         self.run_test(model, x)
 
+    @skipIfUnsupportedMaxOpsetVersion(14)
```
Comment (garymm) [inline test/onnx/test_pytorch_onnx_onnxruntime.py:3563]: I think testing that it successfully exports on opset_version 15 is necessary and sufficient (i.e., replace this test with an opset_version 15 test that doesn't fail).

Comment (garymm) [reply test/onnx/test_pytorch_onnx_onnxruntime.py:3563]: Makes sense. Can you please add some of this to a doc string of the test method?

**Review by garymm — APPROVED** [review]

**Pull request closed without merging**